lang/csharp/src/apache/main/Util/Decimal.cs

```diff
@@ -0,0 +1,177 @@
+
+            var scale = GetLogicalScale(schema);
+
+            if (scale < 0 || scale > precision)
+                throw new AvroTypeException("'decimal' requires a 'scale' property that is zero or less than or equal to 'precision'");
+        }
+
+        /// <summary>
+        /// Converts a logical value to an instance of its base type.
+        /// </summary>
+        /// <param name="logicalValue">The logical value to convert.</param>
+        /// <param name="schema">The schema that represents the target of the conversion.</param>
+        /// <returns>An object representing the encoded value of the base type.</returns>        
+        public override object ConvertToBaseValue(object logicalValue, LogicalSchema schema)
+        {
+            var decimalValue = (decimal)logicalValue;
+            var unscaledInt = new BigInteger(GetUnscaledDecimalValue(decimalValue, out int scale));
+
+            if (scale != GetLogicalScale(schema))
+                throw new ArgumentOutOfRangeException("logicalValue", $"The decimal value has a scale of {scale} which cannot be encoded against a logical 'decimal' with a scale of {GetLogicalScale(schema)}");
+
+            if (decimalValue < 0)
+                unscaledInt = BigInteger.Negate(unscaledInt);
+
+            var buffer = unscaledInt.ToByteArray();
+
+            Array.Reverse(buffer); // Make the sequence of bytes big-endian
+
+            return Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (object)buffer
+                : (object)new GenericFixed((FixedSchema)schema.BaseSchema, buffer);
+        }
+
+        /// <summary>
+        /// Converts a base value to an instance of the logical type.
+        /// </summary>
+        /// <param name="baseValue">The base value to convert.</param>
+        /// <param name="schema">The schema that represents the target of the conversion.</param>
+        /// <returns>An object representing the encoded value of the logical type.</returns>
+        public override object ConvertToLogicalValue(object baseValue, LogicalSchema schema)
+        {
+            var buffer = Schema.Type.Bytes == schema.BaseSchema.Tag
+                ? (byte[])baseValue
+                : ((GenericFixed)baseValue).Value;
+
+            Array.Reverse(buffer); // Make the sequence of bytes little-endian
+
+            var unscaledInt = new BigInteger(buffer);
+            var isNegative = unscaledInt < 0;
+
+            var absUnscaledInt = BigInteger.Abs(unscaledInt);
+
+            buffer = absUnscaledInt.ToByteArray();
+
+            var paddedBuffer = new byte[12];
+
+            Buffer.BlockCopy(buffer, 0, paddedBuffer, 0, Math.Min(buffer.Length, 12));
+
+            var valBits = new int[]
+            {
+                BitConverter.ToInt32(paddedBuffer, 0),
+                BitConverter.ToInt32(paddedBuffer, 4),
+                BitConverter.ToInt32(paddedBuffer, 8)
+            };
```

Comment:
Not 100% sure of this ground, but this is an area I think we have to treat big-endian architectures differently, e.g.:

```
          if (BitConverter.IsLittleEndian)
            {
                // as above
                
            }
            else
            {
                // On a BigEndian architecture, we need the byte order to be bigendian also before
                // interpreting as an int32. However, the array of int32[] passed to the constructor
                // of decimal should still be least significant byte first
                Array.Reverse(paddedBuffer);
                valBits = new int[]
                {
                    BitConverter.ToInt32(paddedBuffer, 8),
                    BitConverter.ToInt32(paddedBuffer, 4),
                    BitConverter.ToInt32(paddedBuffer, 0)                    
                };                
            }
```

Reply:
I think you might be correct. Given where these bytes are now coming from their endianess will need to be considered. Good catch. Will update.